### PR TITLE
dialects: Add a halo dialect [experimental]

### PR DESCRIPTION
**Pull request by AntonLydike**
This is a parking spot for the new halo dialect:

```
halo.swap(%temp) {
  "shape" = #halo.shape_info<[0,4,104,108]x[0,4,104,108]>, 
  "swaps" = [
    #halo.exchange_decl<at [4, 0] size [100, 4] source offset [0, 4] to -1>, 
    #halo.exchange_decl<at [4, 104] size [100, 4] source offset [0, -4]  to 1>
  ],
  "nodes" = 2 : i64
} : (!stencil.temp<[108 : i64, 108 : i64], i32>) -> ()
```

Currently containing:

 - A "fully featured" `halo.swap` op
 - Placeholder `scatter` and `gather` ops to distribute data from one node onto all nodes, and collect the data into a single `memref` afterwards, might not be needed.

### CODE REVIEW

File: xdsl/dialects/experimental/halo.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    """
+    name = "halo.exchange_decl"
+
+    offset_: ParameterDef[builtin.DenseArrayBase]
+    size_: ParameterDef[builtin.DenseArrayBase]
+    source_offset_: ParameterDef[builtin.DenseArrayBase]
+    neighbor_: ParameterDef[builtin.IntegerAttr[builtin.i64]]
+
+    @staticmethod
+    def from_tuples(
```
Comment (PapyChacal) [inline xdsl/dialects/experimental/halo.py:53]: I don't like `from_tuples` given it does not expect tuples specificaly. Maybe something like `from_values`?

File: xdsl/dialects/experimental/halo.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    +--+-----------+--+ c1
+    |  |           |  |
+    +--+-----------+--+ d1
+
+    We can now name these points:
+
+         - a: buffer_start
+         - b: core_start
+         - c: core_end
+         - d: buffer_end
```
Comment (PapyChacal) [inline xdsl/dialects/experimental/halo.py:154]: Are `a` and `d` representing the bounds of the memory buffer, or the halo?
If it supposed to represent the halo, I find the name confusion-prone and would like to suggest renaming it to halo_start/halo_end or similar.

**Review by PapyChacal — APPROVED** [review]
Looks promising!
I have some comments/questions, but this is in experimental, so I'm mostly looking forward for experiments ;)

**Comment by AntonLydike** [comment]
Superseeded by #1005

**Pull request closed without merging**